Mathlib/NumberTheory/Frobenius/Basic.lean

```diff
@@ -0,0 +1,311 @@
+  map_monomial _
+
+lemma F.descent (h : ∀ b : B, (∀ σ : B ≃ₐ[A] B, σ • b = b) → ∃ a : A, b = a) :
+    ∃ m : A[X], (m : B[X]) = F A Q := by
+  choose f hf using h
+  classical
+  let f' : B → A := fun b ↦ if h : ∀ σ : B ≃ₐ[A] B, σ b = b then f b h else 37
+  use (∑ x ∈ (F A Q).support, (monomial x) (f' ((F A Q).coeff x)))
+  ext N
+  push_cast
+  simp_rw [finset_sum_coeff, ← lcoeff_apply, lcoeff_apply, coeff_monomial,
+    Finset.sum_ite_eq', mem_support_iff, ne_eq, ite_not, f']
+  split
+  · next h => exact h.symm
+  · next h1 =>
+    rw [dif_pos <| fun σ ↦ ?_]
+    · refine (hf ?_ ?_).symm
+    · nth_rw 2 [← F.smul_eq_self σ]
+      rfl
+
+-- This will be true in applications to number fields etc. A Galois-invariant element
+-- of `B` is in `A`.
+variable (isGalois : ∀ b : B, (∀ σ : B ≃ₐ[A] B, σ • b = b) → ∃ a : A, b = a)
+
+/-- An auxiliary polynomial, needed in the definition of Frobenius elements.
+It's the descent to `A[X]` of a certain auxiliary Galois-invariant polynomial in `B[X]`. -/
+noncomputable abbrev m := (F.descent A Q isGalois).choose
+
+lemma m_spec : ((m A Q isGalois) : B[X]) = F A Q := (F.descent A Q isGalois).choose_spec
+
+lemma m_spec' : (m A Q isGalois).map (algebraMap A B) = F A Q := by
+  rw [← m_spec A Q isGalois]
+  rfl
+
+-- Amelia's trick to insert "let P be the ideal under Q" into the typeclass system
+variable (P : Ideal A) [P.IsMaximal] [Algebra (A ⧸ P) (B ⧸ Q)] [IsScalarTower A (A⧸P) (B⧸Q)]
+variable {A} -- now can be implicit because `P`
+
+lemma m.y_mod_P_eq_zero : Polynomial.aeval (↑(y A Q) : B ⧸ Q) (m A Q isGalois) = 0 := by
+  rw [← aeval_map_algebraMap B, m_spec', algebraMap.coe_def, aeval_algebraMap_apply,
+    coe_aeval_eq_eval, F.y_eq_zero A Q, map_zero]
+
+open scoped Polynomial
+
+variable [Fintype (A⧸P)]
+
+lemma m.mod_P_y_pow_q_eq_zero' :
+    aeval ((algebraMap B (B⧸Q) (y A Q)) ^ (Fintype.card (A⧸P)))  (m A Q isGalois) = 0 := by
+  let field : Field (A⧸P) := ((Ideal.Quotient.maximal_ideal_iff_isField_quotient P).mp ‹_›).toField
+  rw [← aeval_map_algebraMap (A⧸P), FiniteField.aeval_pow_card, ← algebraMap.coe_def,
+    aeval_map_algebraMap, m.y_mod_P_eq_zero, zero_pow Fintype.card_ne_zero]
+
+lemma F.mod_Q_y_pow_q_eq_zero' :
+    aeval ((algebraMap B (B⧸Q) (y A Q)) ^ (Fintype.card (A⧸P))) (F A Q) = 0 := by
+  rw [← m_spec' A Q isGalois, aeval_map_algebraMap, m.mod_P_y_pow_q_eq_zero']
+
+lemma exists_Frob : ∃ σ : B ≃ₐ[A] B, σ (y A Q) - (y A Q) ^ (Fintype.card (A⧸P)) ∈ Q := by
+  have := F.mod_Q_y_pow_q_eq_zero' Q isGalois P
+  rw [F_spec, aeval_finset_prod, Finset.prod_eq_zero_iff] at this
+  obtain ⟨σ, -, hσ⟩ := this
+  use σ
+  simp only [Ideal.Quotient.algebraMap_eq, AlgEquiv.smul_def, map_sub, aeval_X, aeval_C,
+    sub_eq_zero] at hσ
+  exact (Submodule.Quotient.eq Q).mp (hσ.symm)
+
+/-- An auxiliary arithmetic Frobenius element, in the automorphism group of the integer ring
+rather than the global field itself. -/
+noncomputable abbrev Frob := (exists_Frob Q isGalois P).choose
```

Comment:
The full name is `ArithmeticFrobenius.Abstract.Frob`. I understand you want to make it clear we are talking about the arithmetic one, but `Abstract` seems useless.

Reply:
But we will never use this Frob, as it is an element of Aut(B/A), not Gal(L/K), and one of its inputs is some weird hypothesis about everything in B which is Aut-stable is in A. *Everything* in the Abstract namespace is useless! Actual Frobenius elements are the next PR. As far as I am concerned, this definition could be private!

Reply:
I see, no problem then!.

Reply:
How about I call the namespace Aux?

Reply:
This seems a good idea!